src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs

```diff
@@ -105,14 +110,94 @@ public static class FrozenDictionary
                 // the Equals/GetHashCode methods to be devirtualized and possibly inlined.
                 if (ReferenceEquals(comparer, EqualityComparer<TKey>.Default))
                 {
-                    // In the specific case of Int32 keys, we can optimize further to reduce memory consumption by using
-                    // the underlying FrozenHashtable's Int32 index as the keys themselves, avoiding the need to store the
-                    // same keys yet again.
-                    return typeof(TKey) == typeof(int) ?
-                        (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source) :
-                        new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
+#if NET7_0_OR_GREATER
+                    static FrozenDictionary<TKey, TValue> PickIntegerDictionary<TInt>(Dictionary<TKey, TValue> source)
+                        where TInt : struct, IBinaryInteger<TInt>
+                    {
+                        var k = (TInt[])(object)source.Keys.ToArray();
+                        var v = source.Values.ToArray();
+
+                        Array.Sort(k, v);
+
+                        var min = k[0];
+                        var max = k[k.Length - 1];
+                        var range = ulong.CreateTruncating(max - min);
+
+                        if (range == (ulong)k.Length - 1)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new ClosedRangeIntegerFrozenDictionary<TInt, TValue>(k, v);
```

Comment:
Is this case actually common enough to optimize for it?  Can you share some examples of where it ends up kicking in?  I want us to add optimizations that are valuable, but I also want to avoid maintaining a lot of code that won't actually end up being used.

Once we don't care about the range, then this implementation could be in terms of `IComparable<T>` rather than `IBinaryInteger`, at which point we don't need to special-case .NET 7 vs downlevel.

Reply:
This came up in some general-purpose code that ended up adding a bunch of values from an enum into a set. All these values were contiguous. This can also happen in a dictionary, where you map enum values to something else.

Reply:
I nuked the ClosedRangeInt32/IntegerSet types as these cases are handled nearly as efficiently using the SparseInt32/IntegerSet cases. For the dictionary case, the closed range case remains valuable for enum mapping, among other cases.